### PR TITLE
DAOS-5553 object: memory leak when cleanup DTX handle on client

### PR DESCRIPTION
**Pull request by Nasf-Fan**
The obj_iod_array::oia_iods need to be released when cleanup the
DTX handle on client. For EC object update, recover the original
iod array after reassemble when cleanup related DTX handle.

Signed-off-by: Fan Yong <fan.yong@intel.com>
Co-authored-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Review by mchaarawi — APPROVED** [review]

File: src/object/obj_tx.c [hunk 1]
```diff
@@ -370,24 +370,28 @@ dc_tx_cleanup_one(struct dc_tx *tx, struct daos_cpd_sub_req *dcsr)
 		D_FREE(dcu->dcu_ec_tgts);
 
 		if (reasb_req != NULL) {
-			dcu->dcu_iod_array.oia_iods = reasb_req->orr_uiods;
-			dcsr->dcsr_sgls = reasb_req->orr_usgls;
+			if (reasb_req->orr_uiods != NULL) {
+				dcu->dcu_iod_array.oia_iods =
+							reasb_req->orr_uiods;
+				dcsr->dcsr_sgls = reasb_req->orr_usgls;
```
Comment (liuxuezhao) [inline src/object/obj_tx.c:376]: dcu->dcu_iod_array can reuse iod_array?
I am not familiar with the logic here, just two questions to confirm -
1) why reset oia_iods as orr_uiods (user passed in iod parameter)? and user's IOD cannot be freed right? (below L389 frees it).
2) in dc_tx_classify_update() it assign "dcu->dcu_iod_array.oia_iods = reasb_req->orr_iods",  in that case reasb_req->orr_iods will be freed together with obj_reasb_req_fini (below L379), why free it again in below L389?

Comment (Nasf-Fan) [reply src/object/obj_tx.c:376]: Originally, the "dcu->dcu_iod_array.oia_iods" is allocated via dc_tx_add_update(), it is not user application passed in. So it needs to be freed when cleanup the TX handle on the client. The "dcu->dcu_iod_array.oia_iods" will be reassigned in dc_tx_classify_update(), the old value is save in "reasb_req->orr_uiods". The new value will be auto freed via obj_reasb_req_fini(), but the old value needs to be explicitly freed dc_tx_cleanup_one().

Comment (liuxuezhao) [reply src/object/obj_tx.c:376]: ok, thx

**Review by liuxuezhao — APPROVED** [review]

**Comment by Nasf-Fan** [comment]
@daos-stack/daos-gatekeeper , would you please to help force land this patch, thanks!

**Pull request merged**